### ISSUE TITLE
Experimental model: cross-cutting axis discovery (style-CCA) — a style-aware, wash-diagnostic companion to keyATM

### ISSUE THREAD

**Issue by nealcaren** [nealcaren/topica#608]
## Proposal: cross-cutting axis discovery — a style-aware, "wash"-diagnostic companion to keyATM

*(Reframed from the original emotion-clustering version below. Emotion turned out to be one instance of a more general capability.)*

### The problem

`topica`'s models (LDA, NMF, BERTopic, Top2Vec, keyATM) all partition documents by **content**. That's correct for most tasks, but it means they structurally cannot recover **cross-cutting factors** — emotion, stance, register, speech-act — which vary *within* topics, not across them. keyATM's own motivation is that unsupervised models "fail to measure specific concepts of substantive interest"; this proposal is about the concepts that aren't a coherent *content* cluster at all.

Stress-tested on **GoEmotions** (54k Reddit comments, 27 emotions + neutral): every off-the-shelf model — LDA, NMF, BERTopic, FASTopic, embedding k-means — hits an NMI ≈ 0.10 wall against the emotion labels and only "finds" the 3–4 emotions that *are* their own keyword (`thanks`, `lol`, `love`, `sorry`). **Better embeddings make it worse** (BERTopic/MiniLM produce sharper *content* clusters). Content swamps affect.

### The method: two-view CCA against a content-free style channel

Content is the high-variance nuisance; the cross-cutting factor is low-variance signal riding on top. Partial out content with **no seed words**:

- **View 1** = document content: TF-IDF+LSA (non-embedding, best) *or* a sentence embedding.
- **View 2** = a **content-free style channel**: function-word rates + length-normalized orthography (`!`, `?`, CAPS ratio, elongation `soooo`, emoji, comma rate…).
- **CCA** returns the subspace shared by both views. Because View 2 has *no content*, the shared subspace is the **style-correlated (affect/register/stance)** subspace.

**GoEmotions results** (non-embedding BoW-CCA + over-cluster→merge is best):

| Method | NMI | ARI | purity | notes |
|---|---|---|---|---|
| TF-IDF k-means (baseline) | 0.120 | 0.081 | 0.386 | content clusters |
| embedding (MiniLM) + k-means | 0.065 | 0.016 | 0.330 | worse — sharper content |
| embedding two-view CCA | 0.098 | 0.034 | 0.370 | +51% over raw embed |
| **BoW two-view CCA** | 0.126 | 0.040 | 0.384 | beats embeddings, interpretable |
| **BoW-CCA, over-cluster(50)→merge(28)** | **0.165** | **0.095** | 0.400 | project best; 6 emotions surfaced |

BoW-CCA surfaces the richest emotion partition of any method (gratitude, amusement, love, joy, optimism, remorse at 43–81% purity), and the non-embedding version is cheaper, sklearn-only, and more interpretable than the embedding one.

### It's not an emotion finder — it's a cross-cutting-axis finder

Same pipeline on non-emotional corpora **does not hallucinate emotions** (discriminant validity):

| Corpus | mean cluster emotionality | what it finds instead |
|---|---|---|
| GoEmotions | 0.08–0.13 | gratitude, amusement, joy |
| 20 Newsgroups | **0.00** | speech-act/register: transactional-help vs argumentative; question vs statement; classified-ad vs essay |
| arXiv abstracts | **0.00** | writing register: empirical vs theoretical; expository vs results-reporting |

The canonical correlations stay high on all three — a shared subspace always exists, but **what it contains depends on the register**: affect in emotional text, speech-act in forums, empirical-vs-theoretical in science writing. It degrades *gracefully to topic* rather than inventing structure.

### The no-ground-truth readout (the actual product for applied users)

Most users have no labels, so the model must be **self-describing**. For each canonical axis, report:
1. **canonical-correlation spectrum** — how much separable non-content structure exists and how fast it decays (how many axes to trust);
2. **two-pole × two-view loadings + exemplar docs** — the *style* pole tells you *why* documents group (they're written alike), which distinguishes a genuine register/stance axis from mere topic;
3. **split-half cosine of the axis word-loadings** — is it real or sampling noise (GoEmotions/20NG/arXiv top axes: 0.98 / 0.94 / 0.93).

### Connection to keyATM: "wash" as a validity diagnostic + two wash operators

keyATM anchors a concept with a few seed keywords. But if a seeded topic "measures gratitude" only because `thanks` is present, it counts the *word*, not the *concept*. Test it by **washing the keys** — remove the seed tokens and re-measure recoverability (ROC-AUC of separating the concept):

| Concept | %docs w/ seed | FULL | keys WASHED | STYLE-only | regime |
|---|---|---|---|---|---|
| amusement | 81% | 0.95 | **0.66** | 0.61 | keyword-parroting |
| remorse | 86% | 0.96 | 0.74 | 0.68 | keyword-parroting |
| gratitude | 86% | 0.98 | 0.86 | 0.82 | keyword + real residual |
| admiration | 18% | 0.92 | 0.88 | 0.69 | diffuse content (barely a keyword) |
| anger | 16% | 0.87 | 0.83 | 0.65 | diffuse content |
| curiosity | 10% | 0.78 | 0.77 | **0.91** | style / speech-act |

Two contributions fall out:

1. **The wash-drop `AUC(full) − AUC(washed)` is a validity metric for keyATM.** A big drop (amusement) = keyword-parroting; the "no post-hoc interpretation → objectivity" claim is hollow because you're measuring the seed. A small drop (anger, admiration) = the seeds anchor a real, distributed concept. One cheap ablation, currently reported by nobody.
2. **The style channel is the natural complement to keyATM's content seeds.** Content seeds say *what words* signal a concept; the style view captures *how it's written* — covering keyATM's two blind spots: **diffuse** concepts (anger/admiration are only 16–18% keyword-covered) and **speech-act** concepts (curiosity is best detected by pure style, AUC 0.91, which keyword seeding structurally cannot see).

**Two wash operators, and when each is safe:**

- **Key-wash** (remove a concept's *own* seeds) → diagnostic. Targeted, safe.
- **Topic-wash** (remove the *nuisance* event/topic, keep latent style/affect). Safe only when **targeted** (a few known event terms) or **soft** (residualize against unsupervised content / keep the CCA style-subspace). **Broad word-class deletion fails**: we POS-tagged GoEmotions and washed all nouns to "remove topics" — NMI *fell* 0.099→0.042, because `thanks/love/hate/joke/loss/sorry` are all tagged nouns, i.e. affect and content are **lexically entangled and not separable by any surface word-class**. That's the whole reason the *soft* CCA/residualization separation is necessary — you cannot hand-delete content without deleting affect.

### Proposed `topica` API sketch

```python
from topica import StyleCCA   # working name

m = StyleCCA(
    n_axes=15,                         # canonical dimensions to keep
    content_view="tfidf",              # or "embedding"
    style_features=("function_words", "orthography"),   # content-free view 2
)
m.fit(docs)
m.canonical_corr_        # spectrum: is there separable non-content structure?
m.describe_axis(0)       # style pole / content pole / exemplars (label-free readout)
m.stability_            # split-half cosine per axis
m.topics_               # optional: cluster the axis subspace (over-cluster->merge)

# keyATM-style, with wash diagnostics
m.seed(gratitude=["thank","thanks","appreciate"])
m.wash_drop("gratitude")   # AUC(full) - AUC(keys removed): concept or keyword?
m.topic_wash(seeds=[...])  # targeted removal of a known nuisance topic
```

Fits `topica`'s model interface; new pieces are the content-free style vectorizer, the CCA/whitening step, and the two wash operators — all sklearn-only, no new heavy deps.

### Limitations / open questions

- **Style–topic confound.** Finds the *style-correlated* axis; works best when the cross-cutting factor varies *within* topic (20NG: clean register axis) and blends toward topic when style and topic are locked together (arXiv: field ⇒ style).
- **Inflated top canonical correlation.** Raw CCA on high-dim whitened views overfits the top correlation (GoEmotions axis-1 reads 1.00); needs shrinkage/regularized CCA for honest magnitudes. Directions are stable (split-half 0.98).
- **Length leak** in the style view fixed by rate-normalizing orthographic counts.
- **Open:** POS-strip / affect-tuned encoder as a third view for the hard cognitive emotions; generalize the readout to stance/formality as first-class targets.

### Reproduction

Self-contained `uv` scripts + consolidated `FINDINGS.md` in a scratch repo: `classical.py`, `neural.py`, `bertopic_fastopic.py`, `tier1.py`, `tier2_cca.py`, `tier3_nonembedding.py`, `tier3_merge_discriminant.py`, `axes_explore.py`, `wash_the_keys.py`, `topic_wash.py`. Happy to port `StyleCCA` into a `topica` experimental module.

---

<details><summary>Original proposal (emotion-clustering framing)</summary>

The original version of this issue framed the method purely as an unsupervised emotion-recovery model on GoEmotions. That framing is superseded above: emotion is one instance of the general "dominant style-correlated axis" the method discovers, and the keyATM/wash connection is the sharper contribution.

</details>

**Comment by nealcaren** [comment]
### Applied demo: framing axes in Black-press lynching coverage (+ when topic-wash matters)

Ran the label-free axis readout on a real single-event corpus with no ground truth — 992 lynching-coverage chunks (1905–1929, 33 Black newspapers). We don't care *which* incident each article reports; we want the latent framing/register axis *across* the coverage. The two-pole × two-view + exemplar readout surfaced three interpretable cross-cutting axes:

| Axis | separates |
|---|---|
| 1 | **community-social reportage vs legal-crime reporting** — `mrs/church/rev/meeting/committee` vs `shall/crime/sheriff/county` |
| 2 | **collective advocacy voice vs factual incident report** — `rights/association/congress` + `we/our` vs `rape/charged/killed` + past-tense/digits |
| 3 | **statistical/analytical vs legislative-procedural** — lynching counts (`219 victims, 2,834…`) vs the Dyer Anti-Lynching Bill in the Senate |

i.e. *how* the Black press framed lynching (mobilization / law / advocacy / statistics), orthogonal to which incident — all with no labels.

**The instructive twist: the targeted wash was a near-no-op** (spectrum 0.99→1.00; axes ~identical with and without washing `lynch/mob/negro/…`). This pins down **when topic-wash actually matters:**

- **Multi-topic corpus, latent factor *across* topics** (GoEmotions): content *varies* and dominates variance → must partial it out. **Wash matters.**
- **Single-event corpus, latent framing *within* the event** (this slice): the event vocabulary is near-*constant* across docs → low variance, doesn't discriminate, wasn't structuring the axes anyway → style-CCA surfaces the framing axes directly. **Wash is redundant** (confirmed empirically).

So `topic_wash(targeted=…)` should be documented as a multi-topic-corpus tool; on a mono-event corpus it's a no-op and the plain axis readout suffices.

Caveats reinforced by this run: canonical correlations badly inflated at small n (992 docs vs 120+80-dim views — CCA overfit; directions coherent, magnitudes not — **regularized CCA is needed**); and some axis structure is newspaper masthead/OCR boilerplate vs article body, which a real pipeline must filter first.

**Comment by nealcaren** [comment]
Closing: cross-cutting axis discovery is a topica-original idea with no reference implementation to validate against, so it cannot pass the validation gate for a shipped model. The demos in this thread remain here for reference.
